### PR TITLE
aarch32: Add ability to generate image header

### PR DESCRIPTION
**Pull request by ayankuma**
The image header is compatible for zimage(32) protocol.

Offset  Value          Description
0x24    0x016F2818     Magic number to identify ARM Linux zImage
0x28    start address  The address the zImage starts at
0x2C    end address    The address the zImage ends at

As Zephyr can be built with a fixed load address, Xen/Uboot can read the image header and decide where to copy the Zephyr image.

Signed-off-by: Ayan Kumar Halder <ayan.kumar.halder@amd.com>

### CODE REVIEW

**Comment by ayankuma** [comment]
Hi @carlocaione / @povergoing / Arm experts,

I have created this patch to explain an issue.

Aim :- I want to add an zImage(32) header so that Xen can read the load address from the header and copy the image in the appropriate location.

Issue :- 
When the following 'svc' instruction get executed (as shown from disassembly), 

```
GTEXT(z_arm_cortex_r_svc)
SECTION_FUNC(TEXT, z_arm_cortex_r_svc)
    svc #_SVC_CALL_CONTEXT_SWITCH
3000115c:   ef000000    svc 0x00000000 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<
    bx lr
30001160:   e12fff1e    bx  lr
```

The control jumps to `30000028`  (thinking that this is the `svc` handler).
But now, `30000000` - `30000030` is allocated to the header.
Thus, the new location of svc handler is `30000058`. So the control should jump here.

How do I achieve this ?

From https://developer.arm.com/documentation/100026/0100/system-control/aarch32-register-descriptions/reset-vector-base-address-register , I see that the RVBAR is not accessible in EL1 mode.

**Comment by povergoing** [comment]
It might help if you would like to check with `build/zephyr/zephyr.map`
1) if the image header is at the head of the image, the 0x30000000?
2) if the offset of `_vector_start ` is 0x30000030 as you mentioned?

And also check if the `vbar` has been set to `0x30000030` after this https://github.com/zephyrproject-rtos/zephyr/blob/392a753ac20eb46ba782b5484b2028438f9d9e5e/arch/arm/core/aarch32/prep_c.c#L258

File: arch/arm/core/aarch32/Kconfig [hunk 1]
```diff
@@ -40,6 +40,13 @@ config CPU_AARCH32_CORTEX_R
 	help
 	  This option signifies the use of a CPU of the Cortex-R family.
 
+config AARCH32_IMAGE_HEADER
+	bool "Add image header"
+	default n
```
Comment (stephanosio) [inline arch/arm/core/aarch32/Kconfig:45]: `default n` is redundant.

```suggestion
```

File: arch/arm/core/aarch32/Kconfig [hunk 2]
```diff
@@ -40,6 +40,13 @@ config CPU_AARCH32_CORTEX_R
 	help
 	  This option signifies the use of a CPU of the Cortex-R family.
 
+config AARCH32_IMAGE_HEADER
+	bool "Add image header"
```
Comment (stephanosio) [inline arch/arm/core/aarch32/Kconfig:44]: "Image header" can mean many different things.

Based on the description provided in the commit message, this should be named `ARM_ZIMAGE_HEADER` or similar.

```suggestion
config ARM_ZIMAGE_HEADER
	bool "zImage Header"
```

**Review by stephanosio — DISMISSED** [review]

File: arch/arm/core/aarch32/Kconfig (hunk 1 above)
Comment (ayankuma) [reply arch/arm/core/aarch32/Kconfig]: I have addressed this change.

File: arch/arm/core/aarch32/Kconfig (hunk 2 above)
Comment (ayankuma) [reply arch/arm/core/aarch32/Kconfig]: I have addressed this change.

**Comment by ayankuma** [comment]
@carlocaione @povergoing @stephanosio Many thanks for your feedback. I have incorporated the changes and it can be reviewed.

File: arch/arm/core/aarch32/Kconfig [hunk 3]
```diff
@@ -40,6 +40,12 @@ config CPU_AARCH32_CORTEX_R
 	help
 	  This option signifies the use of a CPU of the Cortex-R family.
 
+config ARM_ZIMAGE_HEADER
+	bool "zImage Header"
+	depends on CPU_AARCH32_CORTEX_R
```
Comment (povergoing) [inline arch/arm/core/aarch32/Kconfig:45]: I don't think `ARM_ZIMAGE_HEADER` should depends on `CPU_AARCH32_CORTEX_R`. The `ARM_ZIMAGE_HEADER` is a common feature, so it should be something like A CPU or A SOC or A board select this feature, what do you think?

Comment (microbuilder) [reply arch/arm/core/aarch32/Kconfig:45]: @ayankuma Any feedback here, since there is a code freeze coming up soon?

Comment (ayankuma) [reply arch/arm/core/aarch32/Kconfig:45]: @povergoing @microbuilder Thanks for the review and reminder, I will fix this and send a new version in few hours.

**Review by carlocaione — DISMISSED** [review]

File: arch/arm/core/aarch32/Kconfig (hunk 3 above)
Comment (ayankuma) [reply arch/arm/core/aarch32/Kconfig]: I have pushed a new patch and fixed this.

**Review by povergoing — DISMISSED** [review]

**Comment by ayankuma** [comment]
> @ayankuma you need to fix the PR to make the CI passing.

@carlocaione @carlescufi Can you remove `DNM` label ? It might make the CI to pass.

```
Pull request is labeled as 'DNM'
[7](https://github.com/zephyrproject-rtos/zephyr/actions/runs/3881820317/jobs/6621254409#step:2:8)
This workflow fails so that the pull request cannot be merged
[8](https://github.com/zephyrproject-rtos/zephyr/actions/runs/3881820317/jobs/6621254409#step:2:9)
Error: Process completed with exit code 1.
```

**Comment by carlocaione** [comment]
> @carlocaione @carlescufi Can you remove `DNM` label ? It might make the CI to pass.

I removed it but you have a lot of build errors that are not dependent on the DNM label.

**Comment by ayankuma** [comment]
> > @carlocaione @carlescufi Can you remove `DNM` label ? It might make the CI to pass.
> 
> I removed it but you have a lot of build errors that are not dependent on the DNM label.

@carlocaione Apologies for bugging you. I see lot of tests are failing with the same reason

zephyr/tests/kernel/sched/metairq/kernel.scheduler.metairq FAILED **unexpected eof** (qemu 0.223s)

Do you know how do I run a specific test (like kernel.scheduler.metairq) ?

**Comment by carlocaione** [comment]
> Then for M class, it generates an incorrect map file

How this can be enabled for M class when the Kconfig depends on `CPU_AARCH32_CORTEX_R || CPU_AARCH32_CORTEX_A`?

For which board / test / sample are you seeing the issue?

**Comment by ayankuma** [comment]
> > Then for M class, it generates an incorrect map file
> 
> How this can be enabled for M class when the Kconfig depends on `CPU_AARCH32_CORTEX_R || CPU_AARCH32_CORTEX_A`?
I have disabled it for M class for the two reasons 

1. Refer arch/arm/core/aarch32/cortex_m/Kconfig , it says that not all the M class will have CPU_CORTEX_M_HAS_VTOR. If any M class does not have VTOR, then it cannot have any header (because the vector table which are at the beginning of an image, needs to be relocated when the image has any header.)
2. If I enable it for M class, it generates an incorrect map file (explained in my previous comment).
> 
> For which board / test / sample are you seeing the issue?
The issue is seen on mps2_an385

**Comment by microbuilder** [comment]
I think it might be the older M0 that doesn't have VTOR (the M0+ does). Anything v7M or v8M should have VTOR, though, but I'll do some digging to see if there are exceptions to that.

**Comment by ayankuma** [comment]
> I think it might be the older M0 that doesn't have VTOR (the M0+ does). Anything v7M or v8M should have VTOR, though, but I'll do some digging to see if there are exceptions to that.
> 
> Edit: `VTOR` is never present on M0 and the M1, both of which have their vector table starting address fixed at `0x00000000`. It is **optional** on Cortex-M0+ and Cortex-M23 processors. This means it should be safe to enable this for ARMv7-M (M3/M4/M7), ARMv8-M Mainline (M33, M55), and ARMv8.1M Mainline (M55). The M23 falls under ARMv7-M Baseline.

Many thanks for looking into this. I can modify the code as follows 
```
config ARM_ZIMAGE_HEADER
       bool "zImage Header"
       depends on CPU_AARCH32_CORTEX_R || CPU_AARCH32_CORTEX_A || CONFIG_CPU_CORTEX_M_HAS_VTOR
       help
         This option adds a zImage Header.

```
Does it look sane to you ?

Also, do you have any inputs on the other issue (ie incorrect map file) ? Can you provide any pointers to debug this.

**Comment by microbuilder** [comment]
> Does it look sane to you ?

Assuming that `HAS_VTOR` is correctly assigned elsewhere it should be fine, yes, but if it isn't correct that's a bug with the soc, board, etc., anyway.
 
> Also, do you have any inputs on the other issue (ie incorrect map file) ? Can you provide any pointers to debug this.

Not familiar with zimage, so I'll need to have a look at the spec first to understand what you're doing here, and then go through the comments. I'll have a look tomorrow.

**Comment by microbuilder** [comment]
@ayankuma Sorry, I've been tied up with the MbedTLS 3.2.1 and TF-M update that we need to get in by Friday. Is this something that's quite important to get into 3.3 before feature freeze this Friday, or can it wait until after 3.3 is tagged for you?

**Comment by ayankuma** [comment]
> @ayankuma Sorry, I've been tied up with the MbedTLS 3.2.1 and TF-M update that we need to get in by Friday. Is this something that's quite important to get into 3.3 before feature freeze this Friday, or can it wait until after 3.3 is tagged for you?

@microbuilder No worries at all. There is no rush for this change. It can wait until 3.3 is tagged.

**Comment by ayankuma** [comment]
a gentle ping on this. :)

**Comment by microbuilder** [comment]
> a gentle ping on this. :)

3.3 release is next week, so once that's out I can cycle back to some of the post 3.3 issues and reviews. Sorry for the delay. :(

**Comment by povergoing** [comment]
> @povergoing Do you have any bandwidth to look at this from the A/R point of view?

Sure, I am going to take a look :)

**Review by povergoing — APPROVED** [review]
Verified on fvp_baser_aemv8r_aarch32

**Review by microbuilder — APPROVED** [review]

**Comment by microbuilder** [comment]
@carlocaione @stephanosio Do you have time to have another look at this?

File: arch/arm/core/aarch32/CMakeLists.txt [hunk 4]
```diff
@@ -34,5 +35,11 @@ add_subdirectory_ifdef(CONFIG_ARM_AARCH32_MMU mmu)
...
 add_subdirectory_ifdef(CONFIG_CPU_AARCH32_CORTEX_A cortex_a_r)
 
+if (CONFIG_ARM_ZIMAGE_HEADER)
+zephyr_linker_sources(ROM_START SORT_KEY 0x0vectors zimage_header.ld)
+zephyr_linker_sources(ROM_START SORT_KEY 0x1vectors vector_table.ld)
+zephyr_linker_sources(ROM_START SORT_KEY 0x2vectors cortex_m/vector_table_pad.ld)
+else()
 zephyr_linker_sources(ROM_START SORT_KEY 0x0vectors vector_table.ld)
 zephyr_linker_sources(ROM_START SORT_KEY 0x1vectors cortex_m/vector_table_pad.ld)
+endif()
```
Comment (carlocaione) [inline arch/arm/core/aarch32/CMakeLists.txt:45]: What about:
```suggestion
if (CONFIG_ARM_ZIMAGE_HEADER)
zephyr_linker_sources(ROM_START SORT_KEY 0x0vectors zimage_header.ld)
endif()
zephyr_linker_sources(ROM_START SORT_KEY 0x1vectors vector_table.ld)
zephyr_linker_sources(ROM_START SORT_KEY 0x2vectors cortex_m/vector_table_pad.ld)
```

Comment (carlescufi) [reply arch/arm/core/aarch32/CMakeLists.txt:45]: @ayankuma ping

Comment (ayankuma) [reply arch/arm/core/aarch32/CMakeLists.txt:45]: @carlescufi , all - Apologies for the delay in response, I was very busy for the last few days. :(
I had tried this suggestion before sending my patch, but it did not work for mps2_an385.

I will try this again this week with the latest Zephyr and let you know.

Comment (carlocaione) [reply arch/arm/core/aarch32/CMakeLists.txt:45]: I mean, it's not an extremely important fix. I'm mostly surprised that this doesn't work to be honest.

Comment (carlescufi) [reply arch/arm/core/aarch32/CMakeLists.txt:45]: @carlocaione should we get this in as-is and then improve it after?

Comment (carlocaione) [reply arch/arm/core/aarch32/CMakeLists.txt:45]: > @carlocaione should we get this in as-is and then improve it after?

I don't want to block this (this is why I commented and not blocked) but I'm a bit worried that the fact that my suggestion is not working is hiding some other kind of problems.

Comment (ayankuma) [reply arch/arm/core/aarch32/CMakeLists.txt:45]: @carlocaione 
I tried your suggestion and tried executing with mps2_an385

With zImage=y, it works fine. The reason being Cortex M expects the first two entries to be "Initial value for the main stack pointer on reset" and "Start address for the reset handler" (Refer Armv8M DDI0553B.vID16122022, B3.30, Vector tables). I have explained this in commit message

 0:   20001938  
 4:   00001185 

With zImage=n, it does **not** work. The first two entries are incorrect.
0:   00000edd   
4:   00000edd 

The reason being with the following
zephyr_linker_sources(ROM_START SORT_KEY 0x1vectors vector_table.ld)
The vector table now shifts to 0x100, so

 100:   20001938
 104:   00001185

Comment (microbuilder) [reply arch/arm/core/aarch32/CMakeLists.txt:45]: I think we can fix that in another PR. Any objections @carlocaione ?

Comment (carlocaione) [reply arch/arm/core/aarch32/CMakeLists.txt:45]: nope, no objections.

Comment (carlescufi) [reply arch/arm/core/aarch32/CMakeLists.txt:45]: @ayankuma merged this one. Please send a follow-up Pull Request with @carlocaione's suggestion.

**Pull request merged**